File: duyo-backend/src/duyo/schemas/language.py
"""Language-practice exercise schemas (mirrors dtm.py's shape)."""

from typing import Literal

from pydantic import BaseModel, Field

from duyo.models.child import AgeSegment


class LanguageQuestion(BaseModel):
    text: str
    choices: list[str]
    correct_index: int
    explanation: str = ""


class LanguagePracticeRequest(BaseModel):
    language: Literal["ru", "en"]
    age_segment: AgeSegment
    topic: str | None = Field(default=None, max_length=60)
    count: int = Field(default=5, ge=1, le=10)


class LanguagePracticeResponse(BaseModel):
    questions: list[LanguageQuestion]

File: duyo-mobile/src/api/endpoints/language.ts
import { AI_TIMEOUT_MS, apiClient } from '@/api/client';
import { type AgeSegment } from '@/api/types';

// Mirrors backend duyo.schemas.language — AI-generated practice exercises,
// grounded in published ContentType.LANGUAGE items when any exist.

export type PracticeLanguage = 'ru' | 'en';

export interface LanguageQuestion {
  text: string;
  choices: string[];
  correct_index: number;
  explanation: string;
}

export interface LanguagePracticeRequest {
  language: PracticeLanguage;
  age_segment: AgeSegment;
  topic?: string;
  count?: number;
}

export interface LanguagePracticeResponse {
  questions: LanguageQuestion[];
}

export async function getLanguageExercises(
  payload: LanguagePracticeRequest,
): Promise<LanguageQuestion[]> {
  const { data } = await apiClient.post<LanguagePracticeResponse>(
    '/language/exercises',
    payload,
    // Generates exercises with Gemini — see api/client.ts on why the 15s
    // default is wrong for anything that waits on a model.
    { timeout: AI_TIMEOUT_MS },
  );
  return data.questions;
}
